new-post: reana-client-python-api (#5)
reanahub/blog.reana.io: content/posts/2021/reana-client-python-api.md

---

> title: How to run workflows using reana-client Python API"

Reviewer: Extra " at the end

> title: How to run workflows using reana-client Python API"

Reviewer 2: ```suggestion
title: How to run workflows using reana-client Python API
```

date: 2021-02-22T15:00:00+01:00
tags:

>   - reana-client

Reviewer: This could perhaps have three tags: API, client, Python.

---

Here you can find a step by step guide on how to submit workflows to
REANA using the `reana-client` Python API. This is a simple example,
for illustrative purposes, using the serial workflow engine and a
simple `ROOT6/RooFit` analysis example.
<!--more-->

Install `reana-client`:
```
$ pip install reana-client==0.7.0
```

And now, inside a Python interpreter, or a Python file:
```python

> # Let us use the REANA example - ROOT6/ RooFit analysis

Reviewer: Can you please break these comments into regular blog post text, so that there would be many blocks of leading text and Python code snippets?  (A more usual blog style.)

Reviewer: ... and the blog post could end with displaying the final graph.

Author: Sure, the reason I initially kept the code in a single block is that now it's very easy to copy the code to a Python script and simply execute it

Reviewer 2: It's a valid point, what about adding a link to a gist at the bottom of the blog post?

> - Access the full Python script here.

or similar.

Reviewer: We could in this case create a Jupyter notebook.  It would have both nice presentation and be fully runnable.

(The advantage of a blog post is that it allows to make detours, to say "what if", to present alternatives... Basically just to teach, not to be blindly copied...)

---
title: "How to run workflows using reana-client Python API"
date: 2021-02-22T15:00:00+01:00

> tags:
>   - API, client, Python

Reviewer 2: ```suggestion
tags:
  - API
  - client
  - Python
```

---
title: "How to run workflows using reana-client Python API"
date: 2021-02-22T15:00:00+01:00
tags:
  - API
  - client
  - Python
---

Here you can find a step by step guide on how to submit workflows to
REANA using the `reana-client` Python API. This is a simple example,
for illustrative purposes, using the serial workflow engine and a
simple [ROOT6 / RooFit analysis](https://github.com/reanahub/reana-demo-root6-roofit/) example.
We will create a new workflow, run it and retrieve its results using
the `reana-client` Python API.
<!--more-->

Install `reana-client`:
```

> $ pip install reana-client==0.7.0

Reviewer 2: Should we point to [0.7.2](https://github.com/reanahub/reana-client/releases/tag/0.7.2)? it seems a bit weird to point to an old version.

```

> Folowing instructions can be executeed inside a Python interpreter,

Reviewer 2: ```suggestion
The following instructions can be executed inside a Python interpreter,
```

or a Python file.

First we need to provide the minimum configuration:

> - Point the REANA service URL:

Reviewer 2: ```suggestion
- Point to the REANA service URL:
```

```python
# with `reana-client` it would be
#   `export REANA_SERVER_URL=https://reana.cern.ch`
import os

> os.environ['REANA_SERVER_URL'] = 'https://reana.cern.ch'

Reviewer: Could be checked first, whether it exists, like `REANA_ACCESS_TOKEN` is, for consistency.

```

- Provide our REANA access token:
```python
# with `reana-client` it would be
#  `export REANA_ACCESS_TOKEN=XXXXXXX`
from getpass import getpass

my_reana_token = \
  os.getenv("REANA_ACCESS_TOKEN") or getpass('Enter your REANA token: ')
```

Now we will choose a name for the workflow and the kind of
workflow specification we will use:
```python
my_workflow_name = 'root6-roofit'

> workflow_type = 'serial'

Reviewer: `my_workflow_type` for consistency.

However, a meta comment. Since most users use Yadage, and since Yadage is actually our default workflow engine in Python API, what about switching the example to use Yadage rather than Serial?

```

> Let us specify inputs and workflows specification:

Reviewer 2: "Let's" instead of "Let us"?

> - So first we will provide it's inputs: (files and parameters)

Reviewer 2: ```suggestion
- So first we will provide its inputs: (files and parameters)
```

Reviewer 2: These two parenthesis blocks separated with the colon look a bit strange to me, maybe we can display the info a bit differently?

Reviewer: Meta-comment: do not use itemised lists, we can use text paragraphs and code blocks. Also, the text paragraphs can be much expanded to use a lot of verbiage, in a tutorial style, so that the text would be highly readable. One example: instead of the very brief "Point the REANA service URL", we can say something very explicit like: "Let us set up the environment variable REANA_SERVER_URL. This variable is mandatory and should point to the remote REANA cluster instance that will be used to run our computations. The client code will use this environment variable automatically to connect to the remote REANA cluster later on. The typical value is https://reana.cern.ch for CERN deployments".  You know what I mean: basically the blog post should read like a small novel :wink: and be fully self-explanatory.

Reviewer: On the same theme: instead of having "with `reana-client` it would be..." as a Python code comment, this text could be moved to the main narrative, expressing it more verbally, for example "This is equivalent to creating the workflow using `reana-client create` command" with link to the docs for the create command.

  (see the [reana.yaml](https://github.com/reanahub/reana-demo-root6-roofit/blob/f29e98b482fe8cb801735ac2fa48bc01e6cc05b7/reana.yaml#L2-L9)
  code that corresponds to it)
```python
my_inputs = {
    'files': [
        'code/gendata.C',
        'code/fitdata.C'
    ],  # A list of files your analysis will be using
    'parameters': {
        'events': '20000',
        'data': 'results/data.root',
        'plot': 'results/plot.png',
    }  # Parameters for your workflow
}
```

- Then we will provide the actual workflow specification,
  in this case using serial workflow:
  (see the [reana.yaml](https://github.com/reanahub/reana-demo-root6-roofit/blob/f29e98b482fe8cb801735ac2fa48bc01e6cc05b7/reana.yaml#L13-L22)
  code that corresponds to it
  )
```python

> workflow_json = {

Reviewer: Name it `my_workflow`, by analogy with `my_inputs`.  The fact that is comes from JSON is not significant.  (And one could ask why we don't call `my_inputs_json` in the former case).  Better to just use my_code, my_inputs, my_workflow, my_results, etc.

    'steps': [
        {'name': 'gendata',
         'environment': 'reanahub/reana-env-root6:6.18.04',
         'commands': [
           'mkdir -p results',
           'root -b -q \'code/gendata.C(${events},"${data}")\' | tee gendata.log']},
        {'name': 'fitdata',
         'environment': 'reanahub/reana-env-root6:6.18.04',
         'commands': [
           'root -b -q \'code/fitdata.C("${data}","${plot}")\' | tee fitdata.log']}]

> }

Reviewer 2: Maybe we can format it with `black`?

```

We are now ready to create our workflow in REANA:
```python
# with `reana-client` it would be
#   `reana-client create -n $my_workflow_name`
from reana_client.api.client import create_workflow_from_json

> create_workflow_from_json(

Reviewer 2: What about adding an empty line to separate the import?

    workflow_json=workflow_json,
    name=my_workflow_name,
    access_token=my_reana_token,
    parameters=my_inputs,
    workflow_engine=workflow_type)
```

Upload files to the workflow workspace:
```python
# with `reana-client` it would be
#   `reana-client upload -w $my_workflow_name`
from reana_client.api.client import upload_to_server

abs_path_to_input_files = \
  [os.path.abspath(f) for f in my_inputs['files']]
upload_to_server(my_workflow_name,
                 abs_path_to_input_files,
                 my_reana_token)
```

> And now we can start our workflow

Reviewer 2: ```suggestion
And now we can start our workflow:
```

```python
# with `reana-client` it would be
#   `reana-client start -w $my_workflow_name`
from reana_client.api.client import start_workflow

start_workflow(my_workflow_name, my_reana_token, {})
```

> We can inquiry for it's status as follows:

Reviewer 2: ```suggestion
We can inquiry for its status as follows:
```

```python
# with `reana-client` it would be
#   `reana-client status -w $my_workflow_name`
import time

while True:
  from reana_client.api.client import get_workflow_status
  status_details = get_workflow_status(my_workflow_name,
                                      my_reana_token)
  if status_details['status'] == 'finished':
    break
  print('Current status: ', status_details['status'])
  time.sleep(1)
```

Once finished, we can list the output files and download
the ones we are interested in:

- Listing files:
```python
# with `reana-client` it would be
#   `reana-client ls -w $my_workflow_name`
from reana_client.api.client import list_files

list_files(my_workflow_name, my_reana_token)
```

- Downloading the result file:
```python
# with `reana-client` it would be
#   `reana-client download results/plot.png -w $my_workflow_name`
from reana_client.api.client import download_file

output_filename = 'results/plot.png'
file_binary_blob = download_file(
  my_workflow_name, output_filename, my_reana_token)
```

> Here you can see the final result:

Reviewer: Add instructions how to plot it, perhaps from Python, perhaps just `os.system()` with `xdg-open`.  The motivation is to provide full instructions so that everything would be runnable.

![sign-up hidden](/images/reana-client-python-api-plot.png)

See also:

> - [The reana-client Python API reference guide](https://reana-client.readthedocs.io/en/latest/#module-reana_client.api.client)

Reviewer: We can link also to the given example.

As noted above, it may be interesting to use Roofit and Yadage perhaps.

---

> title: "How to run workflows using reana-client Python API"

Reviewer: How to run REANA workflows using Python API

(basically, here and elsewhere, we can loose a bit on the term `reana-client`, and just speak of REANA client etc, since people won't be seeing `reana-client` with dash much)

date: 2021-02-22T15:00:00+01:00
tags:
  - API
  - client
  - Python
---

Here you can find a step by step guide on how to submit workflows to
REANA using the `reana-client` Python API. This is a simple example,
for illustrative purposes, using the [Serial](http://docs.reana.io/running-workflows/supported-systems/serial/)
workflow engine and a simple [ROOT6 / RooFit analysis](https://github.com/reanahub/reana-demo-root6-roofit/)
example. We will create a new workflow, run it and retrieve its results
using the `reana-client` Python API.
<!--more-->

### Prerequisite

The only prerequisite to follow this guide is to install `reana-client`, if you
haven't done that already:
```
$ pip install reana-client==0.7.2
```

> ### Configurations

Reviewer: Configuring access

That might be more specific, since "configurations" sounds a bit too generic.

We can actually turn all the headings into this form;

- Installing prerequisites
- Configuring access
- Specifying workflow
- Creating workflow
- Uploading files
- Starting workflow
- Checking workflow status
- Listing workspace files
- Downloading outputs

        'plot': 'results/plot.png',
    }  # Parameters for your workflow
}
```

Finally we need to provide the actual workflow specification where all the
computational steps needed to run our analysis are described. For correspondence
please refer to [reana.yaml](https://github.com/reanahub/reana-demo-root6-roofit/blob/f29e98b482fe8cb801735ac2fa48bc01e6cc05b7/reana.yaml#L13-L22).
The analysis will consist of two stages. In the first stage, signal and
background are generated. In the second stage, a fit will be made for the
signal and background. Here is how we can define the workflow steps:

```python
my_workflow = {
    'steps': [
        {
            'name': 'gendata',
            'environment': 'reanahub/reana-env-root6:6.18.04',
            'commands': [
                'mkdir -p results',
                'root -b -q \'code/gendata.C(${events},"${data}")\' | tee gendata.log',
            ],
        },
        {
            'name': 'fitdata',
            'environment': 'reanahub/reana-env-root6:6.18.04',
            'commands': [
                'root -b -q \'code/fitdata.C("${data}","${plot}")\' | tee fitdata.log'
            ],
        },
    ]
}
```

> ### Workflow creation and usage

Reviewer: Please split "usage" into separate sections (see above), so that we can more easily link to the snippet about getting workflow status, etc.

```

After starting our wokflow we might want to check what is the current status of it.
To inspect the workflow status using `reana-client` we could use [status](https://reana-client.readthedocs.io/en/latest/#reana-client-status)
command. Similarly, it can be achieved with `get_workflow_status` function.
Since it will take couple of seconds for all the computations to be finished
while executing our workflow, we will use a loop to wait and regurlary check
what is the current status of our workflow:

```python
import time

while True:
    from reana_client.api.client import get_workflow_status

    status_details = get_workflow_status(my_workflow_name, my_reana_token)
    if status_details['status'] == 'finished':
        break
    print('Current status: ', status_details['status'])
    time.sleep(1)
```

Once the workflow is finished, we usually want to list all the output files which our workflow
produced. With `reana-client` it can be achieved by using
[ls](https://reana-client.readthedocs.io/en/latest/#reana-client-ls)
command. Similarly, it can be done in the following manner:

```python
from reana_client.api.client import list_files

> list_files(my_workflow_name, my_reana_token)

Reviewer: Meta-comment: It may be useful to also show outputs of each command, in a notebook-like style... In this way, people reading the blog post will be able to picture fully the interactions, what is returned, without actually even running it.

```

After listing all the files we may want to download some of the results for inspection.
With `reana-client` we could use [download](https://reana-client.readthedocs.io/en/latest/#reana-client-download)
command. In this case to download a binary blob of a final plot which our
workflow produced all we need to do is to use `download_file` function:

```python
from reana_client.api.client import download_file

output_filename = 'results/plot.png'
file_binary_blob = download_file(
    my_workflow_name, output_filename, my_reana_token)
```

To display the image we can use Python Imaging Library `Pillow`. If you don't have it
installed in your machine, here is how you can do it:

```

> $ pip install Pillow

Reviewer: Maybe move to the top to Installing requirements?  Saying that we shall use this library to display plots...

Also, you can amend to `pip install --user ...` so that we don't have to explain virtual environment business.

```

To display our plot all we need to so is simply execute the following lines:

```python
from PIL import Image
import io

image_stream = io.BytesIO(file_binary_blob)
img = Image.open(image_stream)
img.show()
```

Here you can see the final result:

![sign-up hidden](/images/reana-client-python-api-plot.png)

See also:
- [The reana-client Python API reference guide](https://reana-client.readthedocs.io/en/latest/#module-reana_client.api.client)

> - [ROOT6 / RooFit analysis example](https://github.com/reanahub/reana-demo-root6-roofit/)

Reviewer: "ROOT6 Roofit" without forward slash

create_workflow_from_json(
    workflow_json=my_workflow,
    name=my_workflow_name,
    access_token=my_reana_token,
    parameters=my_inputs,
    workflow_engine=my_workflow_type)
```

Output:

```python
{
  'message': 'Workflow workspace created',
  'workflow_id': '6cd613eb-f2fb-411b-9601-c89599925759',
  'workflow_name': 'root6-roofit'
}
```

### Uploading files

After creating our workflow we also need to upload all the input files to
the workflow workspace. Same functionality can be achieved with REANA client
[upload](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-upload)
command. It can be done as follows:

```python
from reana_client.api.client import upload_to_server

abs_path_to_input_files = [os.path.abspath(f) for f in my_inputs['files']]
upload_to_server(my_workflow_name, abs_path_to_input_files, my_reana_token)
```

### Starting workflow

At this point we have done all what is needed to start our workflow in REANA.
To do it with REANA client we could call [start](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-start)
command. But in this case it can be started by simply calling `start_workflow`

> function and defining workflow name and access token:

Author: Good catch, I changed to the following:
"But in this case it can be started by simply calling `start_workflow`
function and defining workflow name and access token. Additionally, we
can also override workflow parameters while starting the workflow, but let's
keep the original ones for now:" 
WDYT?

To inspect the workflow status using REANA client we could use [status](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-status)
command. Similarly, it can be achieved with `get_workflow_status` function.
Since it will take couple of seconds for all the computations to be finished
while executing our workflow, we will use a loop to wait and regurlary check
what is the current status of our workflow:

```python
import time
from reana_client.api.client import get_workflow_status

while True:
    status_details = get_workflow_status(my_workflow_name, my_reana_token)
    print('Current status: ', status_details['status'])
    if status_details['status'] == 'finished':
        break
    time.sleep(1)
```

Output:

```python
Current status:  running
...
Current status:  finished
```

### Checking workflow logs

If our workflow has failed or in case we want to debug or investigate how our
workflow works, we may want to display the logs of our workflow. To do that
using REANA client we could use [logs](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-status)
command. In our case the logs can be fetched with `get_workflow_logs` function

> where we only need to specify workflow name and access token. In addition, we

Reviewer 2: ```suggestion
where we only need to specify workflow name and access token. Optionally, we
```

> can add such parameters as a list of `steps` to retrieve logs of only those steps
> which we are interested in or `page` and `size` parameters to paginate the logs

Reviewer 2: ```suggestion
can add such parameters as a list of `steps` whose logs 
we are interested in or `page` and `size` parameters to paginate the logs
```

while True:
    status_details = get_workflow_status(my_workflow_name, my_reana_token)
    print('Current status: ', status_details['status'])
    if status_details['status'] == 'finished':
        break
    time.sleep(1)
```

Output:

```python
Current status:  running
...
Current status:  finished
```

### Checking workflow logs

If our workflow has failed or in case we want to debug or investigate how our
workflow works, we may want to display the logs of our workflow. To do that
using REANA client we could use [logs](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-logs)
command. In our case the logs can be fetched with `get_workflow_logs` function
where we only need to specify workflow name and access token. Optionally, we
can add such parameters as a list of `steps` whose logs we are interested in
or `page` and `size` parameters to paginate the logs output if it gets too
verbose. Workflow logs can be retrieved as follows:

```python
import json
from reana_client.api.client import get_workflow_logs

logs = get_workflow_logs(my_workflow_name, my_reana_token)
logs_json = json.loads(logs['logs'])
print(json.dumps(logs_json, indent=2))
```

Output:

```python
{

>   'workflow_logs': ...,

Reviewer: An example comment: The three dots here make the example a bit opaque. What about doing, as an example, something smaller, notably how to get logs for the second "fitdata" step, and print first say 15 lines of the log?  A little Python loop over steps, localising "gendata" name, and reading logs and spliting by newline and printing first say 15 lines of the "fitdata" step?  Or, even better, print out the last 15 lines of the step, as if seeking for the end of progress... This would make a nice transition to making the plot, because readers will see the plot being generated in the last 15 lines of the fitdata step. I think this would be a rather practical example, where we could even show the exact output, and it would also look better than the json dump in the ipython notebook itself?

command. It can be done as follows:

```python
from reana_client.api.client import upload_to_server

abs_path_to_input_files = [os.path.abspath(f) for f in my_inputs['files']]
upload_to_server(my_workflow_name, abs_path_to_input_files, my_reana_token)
```

### Starting workflow

At this point we have done all what is needed to start our workflow in REANA.
To do it with REANA client we could call [start](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-start)
command. But in this case it can be started by simply calling `start_workflow`
function and defining workflow name and access token. Additionally, we
can also override workflow parameters while starting the workflow, but let's
keep the original ones for now:

```python
from reana_client.api.client import start_workflow

start_workflow(my_workflow_name, my_reana_token, {})
```

Output:

```python
{
  'message': 'Workflow submitted.',
  'run_number': 1,
  'status': 'queued',
  'user': '1d27737b-5b80-4da2-87bb-816a20e110bd',
  'workflow_id': '6cd613eb-f2fb-411b-9601-c89599925759',
  'workflow_name': 'root6-roofit'
}
```

### Checking workflow status

> 

Reviewer: We are now interested to check the workflow status. Is it waiting in the queue, it is running, has it failed ot finished? In command-line interface, we would use the `status` CLI command.. In Python, the corresponding function to use is `get_workflow_status()` function. Since our example workflow may take a few minutes to finish all the computations, we shall use a loop to print its status regularly after a certain wait time:

...

Example output:

...

After starting our workflow we might want to check its current status.
To inspect the workflow status using REANA client we could use [status](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-status)
command. Similarly, it can be achieved with `get_workflow_status` function.
Since it will take couple of seconds for all the computations to be finished
while executing our workflow, we will use a loop to wait and regularly check
what is the current status of our workflow:

```python
import time
from reana_client.api.client import get_workflow_status

while True:
    status_details = get_workflow_status(my_workflow_name, my_reana_token)
    print('Current status: ', status_details['status'])
    if status_details['status'] == 'finished':
        break
    time.sleep(1)
```

Output:

```python
Current status:  running
...
Current status:  finished
```

> ### Checking workflow logs

Reviewer: When our workflow finished, or when we would like to debug and investigate how the workflow run, we may want to display the logs of the concrete computational steps. In command-line interface, we could use the `logs` CLI command. In Python, the corresponding function to use is `get_workflow_logs()` function. We need to specify only the workflow name and the access token. Optionally, we can add parameters such as list of steps the logs of which we are interested to see, or pagination parameters in case the log output is too large or too verbose. As an example, let us retrieve the last 20 lines of the "fitdata" step:

- FIXME or keep the overall "undisplayable" function?

...

Example output:

...

```python
import json
from reana_client.api.client import get_workflow_logs

workflow_logs = get_workflow_logs(my_workflow_name, my_reana_token, ['fitdata'])
job_logs = json.loads(workflow_logs['logs'])['job_logs']
fitdata_logs = job_logs[next(iter(job_logs))]['logs']

for item in fitdata_logs.split('\n')[-15:]:
    print(item)
```

Output:

```python
ERR MATRIX APPROXIMATE
 PARAMETER  CORRELATION COEFFICIENTS
       NO.  GLOBAL      1      2      3      4      5
        1  0.00000   1.000  0.000  0.000  0.000  0.000
        2  0.00000   0.000  1.000  0.000  0.000  0.000
        3  0.00000   0.000  0.000  1.000  0.000  0.000
        4  0.00000   0.000  0.000  0.000  1.000  0.000
        5  0.00000   0.000  0.000  0.000  0.000  1.000
 ERR MATRIX APPROXIMATE
[#1] INFO:Minization -- RooMinimizer::optimizeConst: deactivating const optimization
[#1] INFO:Plotting -- RooAbsPdf::plotOn(model) directly selected PDF components: (bkg)
[#1] INFO:Plotting -- RooAbsPdf::plotOn(model) indirectly selected PDF components: ()
Info in <TCanvas::Print>: png file results/plot.png has been created
```

### Listing workspace files

> 

Reviewer: While the workflow is running or after it finished, we would usually want to list the output files which our workflow produced in its workspace. In command-line interface, we would use the `ls` CLI command. In Python, the corresponding function is `list_files()` function that operates also on the workflow name.

...

Example output:

Once the workflow is finished, we usually want to list all the output files which our workflow
produced. With REANA client it can be achieved by using
[ls](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-ls)
command. Similarly, it can be done in the following manner:

```python
from reana_client.api.client import list_files

list_files(my_workflow_name, my_reana_token)
```

Output:

```python
[
  {
    'last-modified': '2021-02-24T09:57:01',
    'name': 'code/gendata.C',
    'size': 1937,
  },
  {
    'last-modified': '2021-02-24T09:57:01',
    'name': 'code/fitdata.C',
    'size': 1648,
  },
  {
    'last-modified': '2021-02-24T09:57:52',
    'name': 'results/plot.png',
    'size': 15450,
  },
  {
    'last-modified': '2021-02-24T09:57:42',
    'name': 'results/data.root',
    'size': 154458,
  },
]
```

> ### Downloading outputs

Reviewer: We are now interested in downloading workflow outputs, or any temporary files from the workflow's workspace, for closer inspection. In command-line interface, we would use the `download` CLI command. In Python, the corresponding function is `download_file()` function. Let us download a binary blob of the final plot which our workflow produced:

...

After listing all the files we may want to download some of the results for inspection.
With REANA client we could use [download](https://reana-client.readthedocs.io/en/maint-0.7/#reana-client-download)
command. In this case to download a binary blob of a final plot which our
workflow produced all we need to do is to use `download_file` function:

```python
from reana_client.api.client import download_file

output_filename = 'results/plot.png'
file_binary_blob = download_file(
    my_workflow_name, output_filename, my_reana_token)
```

> ### Displaying output images

Reviewer: We can now display the final fit produced by our workflow:

...

Example output:

...
